src/NHibernate/Transform/QueryAliasToObjectPropertySetter.cs

```diff
@@ -0,0 +1,196 @@
+	{
+		private readonly IDictionary<string, NamedMember<FieldInfo>> _fieldsByNameCaseSensitive;
+		private readonly IDictionary<string, NamedMember<FieldInfo>> _fieldsByNameCaseInsensitive;
+		private readonly IDictionary<string, NamedMember<PropertyInfo>> _propertiesByNameCaseSensitive;
+		private readonly IDictionary<string, NamedMember<PropertyInfo>> _propertiesByNameCaseInsensitive;
+
+		private QueryAliasToObjectPropertySetter(IDictionary<string, NamedMember<FieldInfo>> fieldsByNameCaseSensitive,
+			IDictionary<string, NamedMember<FieldInfo>> fieldsByNameCaseInsensitive,
+			IDictionary<string, NamedMember<PropertyInfo>> propertiesByNameCaseSensitive,
+			IDictionary<string, NamedMember<PropertyInfo>> propertiesByNameCaseInsensitive)
+		{
+			_fieldsByNameCaseSensitive = fieldsByNameCaseSensitive;
+			_fieldsByNameCaseInsensitive = fieldsByNameCaseInsensitive;
+			_propertiesByNameCaseSensitive = propertiesByNameCaseSensitive;
+			_propertiesByNameCaseInsensitive = propertiesByNameCaseInsensitive;
+		}
+
+		/// <summary>
+		/// Set the value of a property or field matching an alias.
+		/// </summary>
+		/// <param name="alias">The alias for which resolving the property or field.</param>
+		/// <param name="value">The value to which the property or field should be set.</param>
+		/// <param name="resultObj">The object on which to set the property or field. It must be of the type for which
+		/// this instance has been built.</param>
+		/// <exception cref="PropertyNotFoundException">Thrown if no matching property or field can be found.</exception>
+		/// <exception cref="AmbiguousMatchException">Thrown if many matching properties or fields are found, having the
+		/// same visibility and inheritance depth.</exception>
+		public void SetProperty(string alias, object value, object resultObj)
+		{
+			if (_propertiesByNameCaseSensitive.TryGetValue(alias, out var property))
+			{
+				checkMember(property);
+				property.Member.SetValue(resultObj, value);
+				return;
+			}
+			if (_fieldsByNameCaseSensitive.TryGetValue(alias, out var field))
+			{
+				checkMember(field);
+				field.Member.SetValue(resultObj, value);
+				return;
+			}
+			if (_propertiesByNameCaseInsensitive.TryGetValue(alias, out property))
+			{
+				checkMember(property);
+				property.Member.SetValue(resultObj, value);
+				return;
+			}
+			if (_fieldsByNameCaseInsensitive.TryGetValue(alias, out field))
+			{
+				checkMember(field);
+				field.Member.SetValue(resultObj, value);
+				return;
+			}
+
+			throw new PropertyNotFoundException(resultObj.GetType(), alias, "setter");
+
+			void checkMember<T>(NamedMember<T> member) where T : MemberInfo
+			{
+				if (member.Member != null)
+					return;
+
+				if (member.AmbiguousMembers == null || member.AmbiguousMembers.Length < 2)
+					throw new InvalidOperationException($"{nameof(NamedMember<T>.Member)} missing and {nameof(NamedMember<T>.AmbiguousMembers)} invalid.");
+
+				throw new AmbiguousMatchException(
```

Comment:
What about the `ToPropertiesInsensitivelyDuplicated_WithoutAnyProjections` test I have added with the third commit?